pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,191 @@
+		logger.Infof("Skipping namespace certificate for namespace %s", newNamespace.Name)
+		return nil
+	}
+
+	err = c.reconcile(ctx, newNamespace)
+	if err != nil {
+		c.Recorder.Event(newNamespace, corev1.EventTypeWarning, "InternalError", err.Error())
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error {
+	cfg := config.FromContext(ctx)
+
+	// Only create wildcard certs for the default domain
+	defaultDomain := cfg.Domain.LookupDomainForLabels(nil)
+	if defaultDomain == "" {
+		c.Logger.Infof("Skipping wildcard certificate creation for namespace %s. No default domain defined.", ns.Name)
+		return nil
+	}
+
+	domain, err := getWildcardDomain(cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	if err != nil {
+		c.Logger.Errorf("Failed to apply domain template %s to domain %s and namespace %s", cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	}
+
+	desiredCert := makeDesiredCertificate(ns, []string{domain})
+
+	certs, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	existingCert, err := getExistingCert(domain, certs)
+
+	if apierrs.IsNotFound(err) {
+		cert, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).Create(desiredCert)
+		if err != nil {
+			c.Logger.Errorw("Failed to create namespace certificate", zap.Error(err))
+			c.Recorder.Eventf(ns, corev1.EventTypeWarning, "Creation Failed",
+				"Failed to create Knative certificate %s/%s: %v", ns.Name, desiredCert.ObjectMeta.Name, err)
+			return nil
+		}
+		c.Recorder.Eventf(cert, corev1.EventTypeNormal, "Created",
+			"Created Knative Certificate %s/%s", ns.Name, cert.ObjectMeta.Name)
+	} else if err != nil {
+		return err
+	} else if !equality.Semantic.DeepEqual(existingCert.Spec, desiredCert.Spec) {
+		copy := existingCert.DeepCopy()
+		copy.Spec = desiredCert.Spec
+		_, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(copy.Namespace).Update(copy)
+		if err != nil {
+			c.Logger.Errorw("Failed to update Knative Certificate", zap.Error(err))
+			c.Recorder.Eventf(existingCert, corev1.EventTypeWarning, "UpdateFailed",
+				"Failed to update Knative Certificate %s/%s: %v", desiredCert.Namespace, desiredCert.Name, err)
+			return err
+		}
+		c.Recorder.Eventf(existingCert, corev1.EventTypeNormal, "Updated",
+			"Updated Spec for Knative Certificate %s/%s", desiredCert.Namespace, desiredCert.Name)
+		return nil
+	}
+
+	return nil
+}
+
+func makeDesiredCertificate(namespace *corev1.Namespace, dnsNames []string) *v1alpha1.Certificate {
+	name := wildcardName(namespace.Name)
+	return &v1alpha1.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace.Name,
+		},
+		Spec: v1alpha1.CertificateSpec{
+			DNSNames:   dnsNames,
+			SecretName: name,
+		},
+	}
+}
+
+func getWildcardDomain(tmpl, domain, namespace string) (string, error) {
+	data := network.DomainTemplateValues{
```

Comment:
We need to restrict the format of DomainTemplate here. There is no meaning to create wildcard certificate for  some templates, e.g. `{Namespace}.{Name}.{Domain}`.

The domain with wildcard Name at least needs to have the pattern like `^\*\..+$`